### ISSUE TITLE
Better error handling for failed phone number parsing

### ISSUE THREAD

**Issue by MoritzLost** [rynpsc/craft-phone-number#16]
I just got a nasty bug in production related to phone number formatting, here's the relevant stack trace:

```text
TypeError: libphonenumber\PhoneNumberUtil::format(): Argument #1 ($number) must be of type libphonenumber\PhoneNumber, null given, called in /[...]/vendor/rynpsc/craft-phone-number/src/models/PhoneNumberModel.php on line 106 and defined in /[...]/vendor/giggsey/libphonenumber-for-php/src/PhoneNumberUtil.php:1213
Stack trace:
#0 /[...]/vendor/rynpsc/craft-phone-number/src/models/PhoneNumberModel.php(106): libphonenumber\PhoneNumberUtil->format()
```

The `TypeError` in this case comes from the `format` method, but the underlying cause is this:

```php
        try {
            $this->phoneNumberObject = $this->phoneNumberUtil->parse($number, $region);
        } catch(\Exception $e) {
            // Continue
        }
```

https://github.com/rynpsc/craft-phone-number/blob/master/src/models/PhoneNumberModel.php#L71-L75

If an error occurs for any reason, like the parsing failing somehow, the `phoneNumberObject` property will be null, so calling `format` or any other methods that use it will error out. The try-catch block here is bad style, since it does nothing but suppress the error, causing errors that are hard to debug down the line.

### Suggested solutions

The try-catch block should be removed, allowing the `TypeError` to be thrown if there's an error with parsing the number. This way, it will be noticed early and can be caught more easily. The next step would be to catch that particular error higher up and ensure it causes a validation error so the field can't be saved with a value that will cause this error to happen.

The result should be that as a user of the module, if I call `$entry->getFieldValue('my_phone_number_field')` I will never get a `PhoneNumberModel` in this invalid state – if the value is somehow corrupted, I should get an error right there. Better yet, the field should handle it and return `null` instead.

**Comment by rynpsc** [comment]
As it stands the phone numbers are validated before being saved via the [PhoneNumberValidator](https://github.com/rynpsc/craft-phone-number/blob/master/src/validators/PhoneNumberValidator.php) therefore you should only ever being dealing with valid phone numbers on the frontend so it sounds like the number has been manually edited in the database or something else. I suspect if you were to try and re-save that entry/element in the CP a validation error would be thrown.

Given this this is the only reported issue of anything validation related I am going to assume it's an isolated issue and close. If you are able to consistently replicate the issue please provide further detail.

As a note you should be checking that the field value is not `null` in your templates before calling `format` or any other methods on any field that could return an object or `null`.

```
{% if (entry.phone) %}
  {{ entry.phone.format('national')
{% endif %}

**Comment by MoritzLost** [comment]
@rynpsc I respectfully disagree, both with the premise and the conclusion. If the situation I described is basically impossible – why is the `try...catch` block there at all? I agree that the module doesn't need to handle that edge-case since it's very rare. But if someone _does_ run into this edge-case, all the `try…catch` block does is obfuscate the exception and makes the actual problem harder to track down. So just getting rid of this block would be an instant improvement without any downsides.

**Current Code**

```php
// current code
try {
    $this->phoneNumberObject = $this->phoneNumberUtil->parse($number, $region);
} catch(\Exception $e) {
    // Continue
}
```

**Suggested change**

```php
$this->phoneNumberObject = $this->phoneNumberUtil->parse($number, $region);
```

This way, if something goes wrong in the `parse` method, the request will crash right there and the stack trace will include the actual exception and code that caused the problem. This will make it much easier to debug and fix.

Or is there any reason against this particular change?

----

> it sounds like the number has been manually edited in the database or something else

In this case, the number was added through a migration script. But that script properly created `Entry` objects and validated them – so the number that resulted in this error somehow slipped through the validation. Maybe the number was valid in a previous version of the `libphonenumber` library and became invalid after an update? In any case, due to this I can confidently say that there is at least _one_ edge case where an entry might contain an invalid phone number. By the way, it's also possible to bypass validation before saving an entry, which is necessary in some cases, so there's definitely use-cases where this situation can feasibly occur.

> I suspect if you were to try and re-save that entry/element in the CP a validation error would be thrown.

Yes it would, but until you do that, you still get an exception in the frontend that's difficult to debug, instead of the actual exeption that's causing the error.

> As a note you should be checking that the field value is not null in your templates before calling format or any other methods on any field that could return an object or null.

This doesn't help in this case. If the number is valid, the field won't be empty, so your example code will still result in this error.
